### PR TITLE
Fallback to nodes in the entire cluster, if no nodes are available in specified zones

### PR DESCRIPTION
**Pull request by ashetkar**
- Fallback to nodes in the entire cluster, if no nodes are available in specified zones.
- A new property to override behaviour to fallback to nodes in entire cluster.
- Improved the test.

### CODE REVIEW

File: pgjdbc/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java [hunk 1]
```diff
@@ -243,7 +265,20 @@ public synchronized boolean refresh(Connection conn) throws SQLException {
       return false;
     }
     lastServerListFetchTime = currTime;
-    unreachableHosts.clear();
+    long now = System.currentTimeMillis() / 1000;
+    long failedHostTTL = Long.getLong("failed-host-ttl-seconds", 5);
```
Comment (kneeraj) [inline pgjdbc/src/main/java/com/yugabyte/ysql/ClusterAwareLoadBalancer.java:269]: Please put a comment about why hard coded to 5 seconds. Better to put this in a constant and explain this.

**Review by kneeraj — APPROVED** [review]
Minor comments.

File: pgjdbc/src/test/java/com/yugabyte/FallbackOptionsLBTest.java [hunk 2]
```diff
@@ -5,113 +5,163 @@
...
+  private static final String path = System.getenv("YBDB_PATH");
+  private static String url1 = "jdbc:yugabytedb://localhost:5433/yugabyte?load-balance=true&"
+      + LoadBalanceProperties.TOPOLOGY_AWARE_PROPERTY_KEY + "=";
 
   public static void main(String[] args) throws SQLException, ClassNotFoundException {
-    // Start YugabyteDB cluster if env YBDB_PATH is defined. Else, assume that the cluster is up.
+    if (path == null || path.trim().isEmpty()) {
+      throw new IllegalStateException("YBDB_PATH not defined.");
+    }
+    Class.forName("org.postgresql.Driver");
```
[the remark this answers is not in the document]
Comment (ashetkar) [reply pgjdbc/src/test/java/com/yugabyte/FallbackOptionsLBTest.java:25]: This is smart driver driver itself. Just that the test is run during build and the package relocation (or shading) from `org.postgresql -> com.yugabyte` does not happen during the build process. It happens during packaging.

File: pgjdbc/src/test/java/com/yugabyte/FallbackOptionsLBTest.java [hunk 3]
```diff
@@ -5,113 +5,163 @@
...
+      createConnectionsAndVerify(url, "aws.us-west.us-west-1a", format(-1, -1, -1, 4, 4, 4));
+      createConnectionsAndVerify("jdbc:yugabytedb://127.0.0.4:5433/yugabyte?load-balance=true&explicit-fallback-only=true&topology-keys=",
+          "aws.us-west.us-west-1a", format(-1, -1, -1, 12, 0, 0));
+    } finally {
+      executeCmd(path + "/bin/yb-ctl destroy", "Stop YugabyteDB cluster", 10);
     }
   }
 
-  private static void stopYBDBCluster() {
+  private static void startYBDBClusterWithReadReplica() {
```
[the remark this answers is not in the document]
Comment (ashetkar) [reply pgjdbc/src/test/java/com/yugabyte/FallbackOptionsLBTest.java:92]: No, it's not necessary, any cluster with six nodes would do. Will update the code.

**Pull request merged**